components/raft_log_engine/src/engine.rs

```diff
@@ -658,34 +658,53 @@ impl RaftEngine for RaftLogEngine {
         Ok(())
     }
 
-    fn gc(&self, raft_group_id: u64, from: u64, to: u64) -> Result<usize> {
-        self.batch_gc(vec![RaftLogGcTask {
-            raft_group_id,
-            from,
-            to,
-        }])
+    fn gc(
+        &self,
+        raft_group_id: u64,
+        _from: u64,
+        to: u64,
+        batch: &mut Self::LogBatch,
+    ) -> Result<()> {
+        batch
+            .0
+            .add_command(raft_group_id, Command::Compact { index: to });
+        Ok(())
     }
 
-    fn batch_gc(&self, tasks: Vec<RaftLogGcTask>) -> Result<usize> {
-        let mut batch = self.log_batch(tasks.len());
-        let mut old_first_index = Vec::with_capacity(tasks.len());
-        for task in &tasks {
-            batch
-                .0
-                .add_command(task.raft_group_id, Command::Compact { index: task.to });
-            old_first_index.push(self.0.first_index(task.raft_group_id));
-        }
-
-        self.consume(&mut batch, false)?;
-
-        let mut total = 0;
-        for (old_first_index, task) in old_first_index.iter().zip(tasks) {
-            let new_first_index = self.0.first_index(task.raft_group_id);
-            if let (Some(old), Some(new)) = (old_first_index, new_first_index) {
-                total += new.saturating_sub(*old);
-            }
-        }
-        Ok(total as usize)
+    fn delete_all_states_before(
+        &self,
+        raft_group_id: u64,
+        apply_index: u64,
+        batch: &mut Self::LogBatch,
+    ) -> Result<()> {
+        // Makes sure REGION_STATE_KEY is the smallest of all.
+        debug_assert!(REGION_STATE_KEY < APPLY_STATE_KEY);
+        debug_assert!(REGION_STATE_KEY < FLUSH_STATE_KEY);
+        // And they all have the same length.
+        debug_assert!(REGION_STATE_KEY.len() == APPLY_STATE_KEY.len());
+        debug_assert!(REGION_STATE_KEY.len() == FLUSH_STATE_KEY.len());
+
+        let key_len = REGION_STATE_KEY.len();
+        self.0
+            .scan_raw_messages(
+                raft_group_id,
+                Some(REGION_STATE_KEY),
+                None,
+                false,
```

Comment:
I'm not sure if the performance is the same, but if scan reverse, the logic can be a lot simplified to keep one version before apply index. Or we can also reverse encode the index if forward scan is more performant.